text/2540-formal-function-parameter-attributes.md

```diff
@@ -0,0 +1,581 @@
+
+    ..
+}
+
+fn foo() {
+    // Closures:
+    let bar = |#[orange] x| { .. };
+    let baz = |#[lemon] x: u8, #[orange] y| { .. };
+}
+```
+
+## Trait definitions
+
+An `fn` definition doesn't need to have a body to permit parameter attributes.
+Thus, in `trait` definitions, we may write:
+
+```rust
+trait Beta {
+    fn bar(#[lemon] self, #[orange] x: u8);
+    fn baz(#[lemon] &self, #[orange] x: u8);
+    fn quux(#[lemon] &mut self, #[orange] x: u8);
+}
+```
+
+In Rust 2015, since anonymous parameters are allowed, you may also write:
+
+```rust
+trait Beta {
+    fn bar(#[lemon] self, #[orange] u8); // <-- Note the absence of `x`!
+}
+```
+
+## `fn` types
+
+You can also use attributes in function pointer types.
+For example, you may write:
+
+```rust
+type Foo = fn(#[orange] x: u8);
+type Bar = fn(#[orange] String, #[lemon] y: String);
+```
+
+## Built-in attributes
+
+Attributes attached to formal parameters do not have an inherent meaning in
+the type system or in the language. Instead, the meaning is what your
+procedural macros, the tools you use, or what the compiler interprets certain
+specific attributes as.
+
+As for the built-in attributes and their semantics, we will, for the time being,
+only permit the following attributes on parameters:
+
+- Lint check attributes, that is:
+  `#[allow(C)]`, `#[warn(C)]`, `#[deny(C)]`, `#[forbid(C)]`,
+  and tool lint attributes such as `#[allow(clippy::foobar)]`.
+
+- Conditional compilation attributes:
+
+    - `#[cfg_attr(...)]`, e.g.
+
+      ```rust
+      fn foo(#[cfg_attr(bar, orange)] x: u8) { .. }
+      ```
+
+      If `bar` is active, this is equivalent to:
+
+      ```rust
+      fn foo(#[orange] x: u8) { .. }
+      ```
+
+      And otherwise equivalent to:
+
+      ```rust
+      fn foo(x: u8) { .. }
+      ```
+
+    - `#[cfg(...)]`, e.g.
+
+      ```rust
+      fn foo(#[cfg(bar)] x: u8, y: u16) { .. }
+      ```
+
+      If `bar` is active, this is equivalent to:
+
+      ```rust
+      fn foo(x: u8, y: u16) { .. }
+      ```
+
+      And otherwise equivalent to:
+
+      ```rust
+      fn foo(y: u16) { .. }
+      ```
+
+All other built-in attributes will be rejected with a semantic check.
+For example, you may not write:
+
+```rust
+fn foo(#[inline] bar: u32) { .. }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+Let `OuterAttr` denote the production for an attribute `#[...]`.
+
+On the formal parameters of an `fn` item, including on method receivers,
+and irrespective of whether the `fn` has a body or not, `OuterAttr+` is allowed.
+For example, all the following are valid:
+
+```rust
+fn g1(#[attr1] #[attr2] pat: Type) { .. }
+
+fn g2(#[attr1] x: u8) { .. }
+
+fn g3(#[attr] self) { .. }
+
+fn g4(#[attr] &self) { .. }
+
+fn g5<'a>(#[attr] &mut self) { .. }
+
+fn g6<'a>(#[attr] &'a self) { .. }
+
+fn g7<'a>(#[attr] &'a mut self) { .. }
+
+fn g8(#[attr] self: Self) { .. }
+```
```

Comment:
does this also (need to) cover arbitrary self types?

Reply:
Sure; I've clarified this in https://github.com/rust-lang/rfcs/pull/2565/commits/d4ff5445da83900e2cea8db3c071711985aaed55.